server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -1516,6 +1519,9 @@ public void snapshotShard(Store store, MapperService mapperService, SnapshotId s
                     }
                 }
 
+                // We can skip writing blobs where the metadata hash is equal to the blob's contents because we store the hash/contents
+                // directly in the shard level metadata in this case
+                final boolean needsWrite = md.hash().length != md.length();
```

Comment:
can you assert here that the content is indeed the same? EDIT: happens below. This might be too subtle otherwise. Perhaps also add a method to `StoreFileMetaData` that marks these files specially, so that it's more obvious to identify the files.

Reply:
Done in https://github.com/elastic/elasticsearch/pull/51729/commits/87945413f0095d4674079bf0745a6f3b176fd510 ... actually found a way to make this even safer via the checksum test just in case. Maybe take another look before I merge?